Sources/Apodini/Semantic Model Builder/SharedSemanticModelBuilder.swift

```diff
@@ -118,15 +118,70 @@ class SharedSemanticModelBuilder: SemanticModelBuilder {
                 .whenAllSucceed(guardEventLoopFutures, on: request.eventLoop)
                 .flatMap { _ in
                     request.enterRequestContext(with: component) { component in
-                        var response: Encodable = component.handle()
-                        for responseTransformer in responseModifiers {
-                            response = request.enterRequestContext(with: responseTransformer()) { responseTransformer in
-                                responseTransformer.transform(response: response)
-                            }
+                        let response = component.handle()
+                        let promise = request.eventLoop.makePromise(of: Encodable.self)
+                        let visitor = ActionVisitor(request: request,
+                                                    promise: promise,
+                                                    responseModifiers: responseModifiers)
+                        switch response {
+                        case is ApodiniEncodable:
+                            // is an Action
+                            // use double dispatch to access
+                            // wrapped element
+                            // swiftlint:disable:next force_cast
+                            (response as! ApodiniEncodable).accept(visitor)
+                        default:
+                            // not an action
+                            // we can skip the visitor
+                            visitor.visit(encodable: response)
                         }
-                        return request.eventLoop.makeSucceededFuture(response)
+                        return promise.futureResult
                     }
                 }
         }
     }
 }
+
+struct ActionVisitor: ApodiniEncodableVisitor {
+    var request: Request
+    var promise: EventLoopPromise<Encodable>
+    var responseModifiers: [() -> (AnyResponseTransformer)]
+
+    func visit<Element: Encodable>(encodable: Element) {
+        let result = transformResponse(encodable)
+        promise.succeed(result)
+    }
+
+    func visit<Element: Encodable>(action: Action<Element>) {
+        do {
+            switch action {
+            case let .send(element):
+                let result = transformResponse(element)
+                promise.succeed(result)
+            case let .final(element):
+                let result = transformResponse(element)
+                promise.succeed(result)
+                try request.eventLoop.close()
+            case .end:
+                // send back an empty response
+                promise.succeed("")
+                try request.eventLoop.close()
```

Comment:
Are you sure this is necessary/correct? From my understanding this would close the entire `EventLoop` (which might be handling thousands of requests/connections and not only the one you are handling here). Same applies to L164. If `ActionVisitor.promise` would contain the whole `Action` (that might contain the transformed response), the exporter could handle closing the connection according to their logic.

Reply:
That's a good point. Reading this, I do not think that closing the `EventLoop` is actually the correct thing to do here. I just assumed that closing the `EventLoop` is the intended behavior of the `.end` case.

I removed this.
Instead, the `RESTInterfaceExporter` now sends back an empty message with status `200 OK` (both, for `.nothing` and `.end`). But I am not sure whether that is sufficient for a `.end`?